File: src/terminal/terminal.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Terminal } from './entities/terminal.entity';
import { CreateTerminalDto } from './dto/create-terminal.dto';
import { UpdateTerminalDto } from './dto/update-terminal.dto';

@Injectable()
export class TerminalService {
  constructor(
    @InjectRepository(Terminal)
    private readonly terminalRepository: Repository<Terminal>,
  ) {}

  async create(createTerminalDto: CreateTerminalDto): Promise<Terminal> {
    const terminal = this.terminalRepository.create(createTerminalDto);
    return await this.terminalRepository.save(terminal);
  }

  async findAll(): Promise<Terminal[]> {
    return await this.terminalRepository.find();
  }

  async findOne(id: number): Promise<Terminal> {
    const terminal = await this.terminalRepository.findOne({
      where: { idTerminal: id },
    });
    if (!terminal) {
      throw new NotFoundException(`Terminal con ID ${id} no encontrado`);
    }
    return terminal;
  }

  async update(id: number, updateTerminalDto: UpdateTerminalDto): Promise<Terminal> {
    const terminal = await this.findOne(id);
    Object.assign(terminal, updateTerminalDto);
    return await this.terminalRepository.save(terminal);
  }

  async remove(id: number): Promise<void> {
    const terminal = await this.findOne(id);
    await this.terminalRepository.remove(terminal);
  }
}

File: src/main/java/com/tallermecanico/service/TerminalService.java
package com.tallermecanico.service;

import com.tallermecanico.dto.request.CreateTerminalDto;
import com.tallermecanico.dto.request.UpdateTerminalDto;
import com.tallermecanico.entity.Terminal;
import com.tallermecanico.exception.DuplicateResourceException;
import com.tallermecanico.exception.ResourceNotFoundException;
import com.tallermecanico.repository.TerminalRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class TerminalService {

    private final TerminalRepository terminalRepository;

    public TerminalService(TerminalRepository terminalRepository) {
        this.terminalRepository = terminalRepository;
    }

    @Transactional
    public Terminal create(CreateTerminalDto dto) {
        // Validar terminal único
        if (dto.getTerminal() != null && terminalRepository.existsByTerminal(dto.getTerminal())) {
            throw new DuplicateResourceException("Terminal", "terminal", dto.getTerminal());
        }

        // Validar prefijo único
        if (dto.getPrefijo() != null && terminalRepository.existsByPrefijo(dto.getPrefijo())) {
            throw new DuplicateResourceException("Terminal", "prefijo", dto.getPrefijo());
        }

        Terminal terminal = new Terminal();
        terminal.setTerminal(dto.getTerminal());
        terminal.setPrefijo(dto.getPrefijo());

        return terminalRepository.save(terminal);
    }

    @Transactional(readOnly = true)
    public List<Terminal> findAll() {
        return terminalRepository.findAll();
    }

    @Transactional(readOnly = true)
    public Terminal findById(Integer id) {
        return terminalRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Terminal", "id", id));
    }

    @Transactional
    public Terminal update(Integer id, UpdateTerminalDto dto) {
        Terminal terminal = findById(id);

        // Validar terminal único (excluyendo el actual)
        if (dto.getTerminal() != null &&
            terminalRepository.existsByTerminalAndIdTerminalNot(dto.getTerminal(), id)) {
            throw new DuplicateResourceException("Terminal", "terminal", dto.getTerminal());
        }

        // Validar prefijo único (excluyendo el actual)
        if (dto.getPrefijo() != null &&
            terminalRepository.existsByPrefijoAndIdTerminalNot(dto.getPrefijo(), id)) {
            throw new DuplicateResourceException("Terminal", "prefijo", dto.getPrefijo());
        }

        if (dto.getTerminal() != null) {
            terminal.setTerminal(dto.getTerminal());
        }
        if (dto.getPrefijo() != null) {
            terminal.setPrefijo(dto.getPrefijo());
        }

        return terminalRepository.save(terminal);
    }

    @Transactional
    public void delete(Integer id) {
        Terminal terminal = findById(id);
        terminalRepository.delete(terminal);
    }
}
